joblib/numpy_pickle.py

```diff
@@ -440,33 +472,49 @@ def load(filename, mmap_mode=None):
     object might not match the original pickled object. Note that if the
     file was saved with compression, the arrays cannot be memmaped.
     """
-    with open(filename, 'rb') as file_handle:
-        # We are careful to open the file handle early and keep it open to
-        # avoid race-conditions on renames. That said, if data are stored in
-        # companion files, moving the directory will create a race when
-        # joblib tries to access the companion files.
-        if _read_magic(file_handle) == _ZFILE_PREFIX:
-            if mmap_mode is not None:
-                warnings.warn('file "%(filename)s" appears to be a zip, '
-                              'ignoring mmap_mode "%(mmap_mode)s" flag passed'
-                              % locals(), Warning, stacklevel=2)
-            unpickler = ZipNumpyUnpickler(filename, file_handle=file_handle)
-        else:
-            unpickler = NumpyUnpickler(filename, file_handle=file_handle,
-                                       mmap_mode=mmap_mode)
-
-        try:
-            obj = unpickler.load()
-        except UnicodeDecodeError as exc:
-            # More user-friendly error message
-            if PY3_OR_LATER:
-                new_exc = ValueError(
-                    'You may be trying to read with '
-                    'python 3 a joblib pickle generated with python 2. '
-                    'This feature is not supported by joblib.')
-                new_exc.__cause__ = exc
-                raise new_exc
-        finally:
-            if hasattr(unpickler, 'file_handle'):
-                unpickler.file_handle.close()
-        return obj
+    # Backward compatibility with old compression strategy
+    with open(filename, 'rb') as f:
+        compressor = _detect_compressor(f)
+        if compressor == 'compat':
+            warnings.warn("The file '%s' has been generated with a joblib "
+                          "version less than 0.10. "
+                          "Please regenerate this pickle file." % filename,
+                          DeprecationWarning, stacklevel=2)
+            return load_compatibility(filename)
+
+        with _read_fileobject(f, compressor) as f:
+            if compressor in _COMPRESSORS and mmap_mode is not None:
+                warnings.warn('File "%(filename)s" is compressed using '
+                              '"%(compressor)s" which is not compatible with '
+                              'mmap_mode "%(mmap_mode)s" flag passed.'
+                              % locals(),
+                              DeprecationWarning, stacklevel=2)
```

Comment:
> > I could make it work (at least on Linux) but has to use a contextlib
> > decorator because of the load_compatibility function. I hope this is fine.
> 
> I haven't looked in detail why you need a contextmanager

It's because the all is done now in the `_read_fileobject` function, including the compatibility check. The `load_compatibility` function directly returns the loaded object, I have to call it in the main load code.

> but I don't think my original comment was addressed:
> 
> > I am wondering whether we could not put the _detect_compressor +
> > _read_fileobject in a single function
> 
> Basically it sounds a bit convoluted to have _detect_compressor return a
> string to say which compressed file we are dealing with, and then
> _read_fileobject take this string and return a fileobj. Why not have only
> one function and return a fileobj directly? You are not using
> _detect_compressor anywhere else as far as I can see.

It's used in some test to check the correct compressor is used. Now this function is only used as an internal helper but not from the main `load` function.